closed/src/java.base/share/classes/com/ibm/security/criu/DigestBase.java

```diff
@@ -0,0 +1,282 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * ===========================================================================
+ */
+
+package com.ibm.security.criu;
+
+import java.security.MessageDigestSpi;
+import java.security.DigestException;
+import java.security.ProviderException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.ArrayList;
+
+import sun.security.action.GetPropertyAction;
+
+/**
+ * Common base message digest implementation for the Sun provider.
+ * It implements all the JCA methods as suitable for a Java message digest
+ * implementation of an algorithm based on a compression function (as all
+ * commonly used algorithms are). The individual digest subclasses only need to
+ * implement the following methods:
+ *
+ *  . abstract void implCompress(byte[] b, int ofs);
+ *  . abstract void implDigest(byte[] out, int ofs);
+ *  . abstract void implReset();
+ *
+ * See the inline documentation for details.
+ *
+ * @since   1.5
+ * @author  Andreas Sterbenz
+ */
+abstract class DigestBase extends MessageDigestSpi implements Cloneable {
+
+    // one element byte array, temporary storage for update(byte)
+    private byte[] oneByte;
+
+    // algorithm name to use in the exception message
+    private final String algorithm;
+    // length of the message digest in bytes
+    private final int digestLength;
+
+    // size of the input to the compression function in bytes
+    private final int blockSize;
+    // buffer to store partial blocks, blockSize bytes large
+    // Subclasses should not access this array directly except possibly in their
+    // implDigest() method. See MD5.java as an example.
+    byte[] buffer;
+    // offset into buffer
+    private int bufferOffset;
+
+    // number of bytes processed so far
+    // subclasses should not modify this value
+    // also used as a flag to indicate reset status
+    // -1: need to call engineReset() before next call to update()
+    //  0: is already reset
+    long bytesProcessed;
+
+    private static final ArrayList<DigestBase> digestList = new ArrayList<>();
+
+    // tracing for CRIUSEC
+    private static final boolean debug = Boolean.parseBoolean(GetPropertyAction.privilegedGetProperty
+            ("enable.j9internal.checkpoint.security.api.debug", "false"));
+
+    /**
+     * This method is required to ensure that no state is maintained when checkpointing.
+     */
+    static void resetDigests() {
+        if (debug) {
+            System.out.println("Clearing all digests ...");
+        }
+        synchronized (digestList) {
+            for (DigestBase digest : digestList) {
+                digest.engineReset();
+            }
```

Comment:
Should we have `digestList.clear();` here? The list of digest objects could be quite large.

Reply:
added it, thank you